sklearn/neighbors/nca.py

```diff
@@ -0,0 +1,545 @@
+                    .format(init.shape[0], init.shape[1]))
+
+            if self.n_features_out is not None:
+                # Assert that self.n_features_out = init.shape[0]
+                if self.n_features_out != init.shape[0]:
+                    raise ValueError(
+                        'The preferred embedding dimensionality '
+                        '`n_features_out` ({}) does not match '
+                        'the output dimensionality of the given '
+                        'linear transformation `init` ({})!'
+                        .format(self.n_features_out,
+                                init.shape[0]))
+
+        elif init in ['pca', 'identity', 'random']:
+            pass
+        else:
+            raise ValueError(
+                "`init` must be 'pca', 'identity', 'random' or a numpy "
+                "array of shape (n_features_out, n_features).")
+
+        return X_valid, y_valid, init
+
+    def _initialize(self, X, init):
+        """Initialize the transformation.
+
+        Parameters
+        ----------
+        X : array-like, shape (n_samples, n_features)
+            Data samples.
+
+        init : string or numpy array of shape (n_features_a, n_features_b)
+            The validated initialization of the linear transformation.
+
+        Returns
+        -------
+        transformation : array, shape (n_features_out, n_features)
+            The initialized linear transformation.
+
+        """
+
+        transformation = init
+
+        if isinstance(init, np.ndarray):
+            pass
+        else:
+            n_features_out = self.n_features_out or X.shape[1]
+            if init == 'identity':
+                transformation = np.eye(n_features_out, X.shape[1])
+            elif init == 'random':
+                transformation = self.random_state_.randn(n_features_out,
+                                                          X.shape[1])
+            elif init == 'pca':
+                pca = PCA(n_components=n_features_out,
+                          random_state=self.random_state_)
+                t_pca = time.time()
+                if self.verbose:
+                    print('Finding principal components... ', end='')
+                    sys.stdout.flush()
+
+                pca.fit(X)
+                if self.verbose:
+                    print('done in {:5.2f}s'.format(time.time() - t_pca))
+
+                transformation = pca.components_
+        return transformation
+
+    def _callback(self, transformation):
+        """Called after each iteration of the optimizer.
+
+        Parameters
+        ----------
+        transformation : array, shape(n_features_out, n_features)
+            The solution computed by the optimizer in this iteration.
+        """
+        if self.callback is not None:
+            self.callback(transformation, self.n_iter_)
+
+        self.n_iter_ += 1
+
+    def _loss_grad_lbfgs(self, transformation, X, y, diffs,
```

Comment:
one can use a `sign` argument when maximizing a function, maybe more elegant than returning - the loss and the gradient:
https://docs.scipy.org/doc/scipy/reference/tutorial/optimize.html#constrained-minimization-of-multivariate-scalar-functions-minimize

also when verbose is activated, is the objective value printed in negative or positive form? it would be best to print in positive form (as it is the original objective function of the paper).

Reply:
Done